File: apps/api/app/services/phone_e164.py
from __future__ import annotations


def normalize_e164(raw: str) -> str:
    """Normaliza a E.164 con prefijo +. Acepta dígitos sin + (p. ej. móvil PE 9 dígitos → +51…)."""
    s = (raw or "").strip().replace(" ", "").replace("-", "")
    if not s.startswith("+"):
        digits = "".join(c for c in s if c.isdigit())
        if not digits:
            msg = "Indica un número de teléfono válido (con o sin +, ej. +51999888777 o 999888777)."
            raise ValueError(msg)
        # Perú: móvil 9 dígitos que empieza por 9
        if len(digits) == 9 and digits[0] == "9":
            s = "+51" + digits
        elif 10 <= len(digits) <= 15:
            s = "+" + digits
        else:
            msg = (
                "Número incompleto o no reconocido. Usa formato internacional con código de país "
                "(ej. +51999888777) o 9 dígitos de móvil peruano (9…)."
            )
            raise ValueError(msg)
    rest = s[1:]
    if not rest.isdigit() or len(rest) < 8 or len(rest) > 18:
        msg = "Número de teléfono inválido (revisa el código de país y la cantidad de dígitos)."
        raise ValueError(msg)
    return s

File: apps/web/src/lib/phone-e164.ts
/**
 * Alineado con `apps/api/app/services/phone_e164.normalize_e164` para que el panel
 * envíe el mismo E.164 que validará el backend.
 */
export function normalizePanelPhoneE164(raw: string): string {
  let s = (raw ?? "").trim().replace(/\s/g, "").replace(/-/g, "");
  if (!s.startsWith("+")) {
    const digits = s.replace(/\D/g, "");
    if (!digits) {
      throw new Error(
        "Indica un número válido (con o sin +, ej. +51999888777 o 999888777).",
      );
    }
    if (digits.length === 9 && digits[0] === "9") {
      s = `+51${digits}`;
    } else if (digits.length >= 10 && digits.length <= 15) {
      s = `+${digits}`;
    } else {
      throw new Error(
        "Número incompleto o no reconocido. Usa código de país (ej. +51999888777) o 9 dígitos de móvil peruano (9…).",
      );
    }
  }
  const rest = s.slice(1);
  if (!/^\d+$/.test(rest) || rest.length < 8 || rest.length > 18) {
    throw new Error("Número de teléfono inválido (revisa el código de país y los dígitos).");
  }
  return s;
}
